src/components/Comment/index.spec.tsx

```diff
@@ -495,103 +518,220 @@ describe(__filename, () => {
   });
 
   describe('discarding a new comment', () => {
-    const getKeyParams = () => {
-      return { versionId: nextUniqueId(), fileName: 'example.json', line: 100 };
-    };
-
-    it('renders a discard button to cancel entering a comment', () => {
+    const setUpForDiscard = (commentId: undefined | number) => {
       const store = configureStore();
-      const keyParams = getKeyParams();
-      store.dispatch(actions.beginComment(keyParams));
-      const dispatchSpy = spyOn(store, 'dispatch');
+      const versionId = nextUniqueId();
 
-      const root = render({
+      if (commentId) {
+        dispatchComment({
+          comment: createFakeExternalComment({ id: commentId }),
+          store,
+          versionId,
+        });
+      }
+
+      return {
+        keyParams: {
+          commentId,
+          fileName: 'example.json',
+          line: 100,
+          versionId,
+        },
         store,
-        ...keyParams,
-      });
-      const discardButton = root.find(`.${styles.discardButton}`);
+      };
+    };
 
-      expect(discardButton).toHaveText('Discard');
+    it.each([undefined, 33])(
+      'renders a discard button to cancel entering a comment with id=%s',
+      (commentId) => {
+        const { keyParams, store } = setUpForDiscard(commentId);
+        store.dispatch(actions.beginComment(keyParams));
+        const dispatchSpy = spyOn(store, 'dispatch');
 
-      discardButton.simulate('click');
+        const root = render({
+          store,
+          ...keyParams,
+        });
+        const discardButton = root.find(`.${styles.discardButton}`);
 
-      expect(dispatchSpy).toHaveBeenCalledWith(
-        actions.considerDiscardComment(keyParams),
-      );
-    });
+        expect(discardButton).toHaveText('Discard');
 
-    it('renders confirm button to discard a comment', () => {
-      const store = configureStore();
-      const keyParams = getKeyParams();
-      store.dispatch(actions.beginComment(keyParams));
-      store.dispatch(actions.considerDiscardComment(keyParams));
-      const dispatchSpy = spyOn(store, 'dispatch');
+        discardButton.simulate('click');
 
-      const root = render({
-        store,
-        ...keyParams,
-      });
-      const discardButton = root.find(`.${styles.discardButton}`);
+        expect(dispatchSpy).toHaveBeenCalledWith(
+          actions.considerDiscardComment(keyParams),
+        );
+      },
+    );
 
-      expect(discardButton).toHaveText('Confirm discard');
+    it.each([undefined, 33])(
```

Comment:
Multiplying all of these tests by two to cover both the number / undefined cases is not really ideal because it duplicates some test logic and that causes slowness. However, I think it's only minimal duplication so it's fine. 

Just as a point of reference, cases like this suggest that the logic needs to be pushed into another *unit* of abstraction so that the variation (number / undefined) can be tested in the most isolated way possible.

Here is an example of what I mean (but I don't apply this patch):

<details>
<summary>Patch</summary>

```diff
diff --git a/src/components/Comment/index.spec.tsx b/src/components/Comment/index.spec.tsx
index 298c2fe..c7fd34c 100644
--- a/src/components/Comment/index.spec.tsx
+++ b/src/components/Comment/index.spec.tsx
@@ -564,24 +564,53 @@ describe(__filename, () => {
       },
     );
 
+    it('renders confirm button to discard a comment', () => {
+      const { keyParams, store } = setUpForDiscard(undefined);
+      store.dispatch(actions.beginComment(keyParams));
+      store.dispatch(actions.considerDiscardComment(keyParams));
+
+      const root = render({
+        store,
+        ...keyParams,
+      });
+      const onDiscard = jest.spyOn(root.instance() as CommentBase, 'onDiscard');
+      const discardButton = root.find(`.${styles.discardButton}`);
+
+      expect(discardButton).toHaveText('Confirm discard');
+
+      discardButton.simulate('click');
+
+      // As you can see, it no longer matters how this is called, as long
+      // as it's called.
+      expect(onDiscard).toHaveBeenCalled();
+    });
+
     it.each([undefined, 33])(
-      'renders confirm button to discard a comment with id=%s',
+      'dispatches finishComment() with the correct parameters, including comment=%s',
       (commentId) => {
-        const { keyParams, store } = setUpForDiscard(commentId);
-        store.dispatch(actions.beginComment(keyParams));
-        store.dispatch(actions.considerDiscardComment(keyParams));
+        const store = configureStore();
         const dispatchSpy = spyOn(store, 'dispatch');
+        const versionId = nextUniqueId();
+        const keyParams = {
+          commentId,
+          fileName: 'manifest.json',
+          line: nextUniqueId(),
+          versionId,
+        };
 
-        const root = render({
-          store,
-          ...keyParams,
-        });
-        const discardButton = root.find(`.${styles.discardButton}`);
-
-        expect(discardButton).toHaveText('Confirm discard');
+        if (commentId) {
+          dispatchComment({
+            comment: createFakeExternalComment({ id: commentId }),
+            store,
+            versionId,
+          });
+        }
+        const root = render({ store, ...keyParams });
 
-        discardButton.simulate('click');
+        (root.instance() as CommentBase).onDiscard();
 
+        // Now we can test the dispatching logic in isolation, including
+        // all of the possible variations.
         expect(dispatchSpy).toHaveBeenCalledWith(
           actions.finishComment(keyParams),
         );

```

</details>

The idea above is that the unit of abstraction is `onDiscard()` so you can 1) pass responsibility to that function by asserting it gets called and 2) execute that function directly to test all of the variations needed.

As you can see, this approach doesn't reduce much duplication (just some of the dispatches) so it's not worth it. However, in some future cases the duplication may be more pronounced and this strategy might be a good one.

To reiterate, your approach here is fine 😄

Reply:
Thanks for showing me this. I updated the patch.